src/Juvix/Compiler/Tree/EvaluatorEff.hs

```diff
@@ -0,0 +1,414 @@
+          cl' <- eval' _nodeCallClosuresFun
+          go cl' vs
+          where
+            go :: Value -> [Value] -> Eff r' Value
+            go cl vs = case cl of
+              ValClosure Closure {..}
+                | argsNum == n -> do
+                    let ctx' =
+                          EvalCtx
+                            { _evalCtxArgs = vs',
+                              _evalCtxTemp = mempty
+                            }
+                    withCtx ctx' (eval' body)
+                | argsNum < n -> do
+                    let ctx' =
+                          EvalCtx
+                            { _evalCtxArgs = take argsNum vs',
+                              _evalCtxTemp = mempty
+                            }
+
+                    body' <- withCtx ctx' (eval' body)
+                    go body' (drop argsNum vs')
+                | otherwise ->
+                    return
+                      ( ValClosure
+                          Closure
+                            { _closureSymbol,
+                              _closureArgs = vs'
+                            }
+                      )
+                where
+                  fi = lookupFunInfo tab _closureSymbol
+                  argsNum = fi ^. functionArgsNum
+                  vs' = _closureArgs ++ vs
+                  n = length vs'
+                  body = fi ^. functionCode
+              _ ->
+                evalError "expected a closure"
+
+        goBranch :: NodeBranch -> Eff r' Value
+        goBranch NodeBranch {..} = do
+          arg' <- eval' _nodeBranchArg
+          br <- case arg' of
+            ValBool True -> return _nodeBranchTrue
+            ValBool False -> return _nodeBranchFalse
+            _ -> evalError "expected a boolean"
+          eval' br
+
+        goCase :: NodeCase -> Eff r' Value
+        goCase NodeCase {..} = do
+          arg' <- eval' _nodeCaseArg
+          case arg' of
+            v@(ValConstr Constr {..}) ->
+              case find (\CaseBranch {..} -> _caseBranchTag == _constrTag) _nodeCaseBranches of
+                Just CaseBranch {..} -> goCaseBranch v _caseBranchSave _caseBranchBody
+                Nothing -> do
+                  def <- maybe (evalError "no matching branch") return _nodeCaseDefault
+                  goCaseBranch v False def
+            _ ->
+              evalError "expected a constructor"
+
+        withExtendedTemp :: Value -> Eff r' a -> Eff r' a
+        withExtendedTemp v m = do
+          ctx <- E.ask
+          withCtx (over evalCtxTemp (BL.cons v) ctx) m
+
+        goCaseBranch :: Value -> Bool -> Node -> Eff r' Value
+        goCaseBranch v bSave body
+          | bSave = withExtendedTemp v (eval' body)
+          | otherwise = eval' body
+
+        goSave :: NodeSave -> Eff r' Value
+        goSave NodeSave {..} = do
+          v <- eval' _nodeSaveArg
+          withExtendedTemp v (eval' _nodeSaveBody)
+
+printValue :: InfoTable -> Value -> Text
```

Comment:
There is code duplication here. Exactly the same function is available in Evaluator.hs.

Reply:
done